majortom_gateway/gateway_api.py

```diff
@@ -0,0 +1,375 @@
+        self.host = host
+        self.gateway_token = gateway_token
+        self.ssl_verify = ssl_verify
+        self.basic_auth = basic_auth
+        self.https = https
+        if ssl_verify is True and ssl_ca_bundle is None:
+            raise(ValueError('"ssl_ca_bundle" must be a valid path to a certificate bundle if "ssl_verify" is True. Could fetch from https://curl.haxx.se/docs/caextract.html'))
+        else:
+            self.ssl_ca_bundle = ssl_ca_bundle
+        self.__build_endpoints()
+        self.command_callback = command_callback
+        self.error_callback = error_callback
+        self.rate_limit_callback = rate_limit_callback
+        self.cancel_callback = cancel_callback
+        self.websocket = None
+        self.queued_payloads = []
+        self.headers = {
+            "X-Gateway-Token": self.gateway_token
+        }
+        if self.basic_auth != None:
+            userAndPass = b64encode(str.encode(f"{self.basic_auth}")).decode("ascii")
+            self.headers['Authorization'] = f'Basic {userAndPass}'
+
+    def __build_endpoints(self):
+        if self.https:
+            self.gateway_endpoint = "wss://" + self.host + "/gateway_api/v1.0"
+        else:
+            self.gateway_endpoint = "ws://" + self.host + "/gateway_api/v1.0"
+
+    async def connect(self):
+        if self.https:
+            ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+
+            if self.ssl_verify:
+                ssl_context.verify_mode = ssl.CERT_REQUIRED
+                ssl_context.check_hostname = True
+                # Should probably fetch from https://curl.haxx.se/docs/caextract.html
+                ssl_context.load_verify_locations(self.ssl_ca_bundle)
+            else:
+                ssl_context.check_hostname = False
+                ssl_context.verify_mode = ssl.CERT_NONE
+
+        else:
+            ssl_context = None
+
+        logger.info("Connecting to Major Tom")
+        try:
+            websocket = await websockets.connect(self.gateway_endpoint,
+                                                 extra_headers=self.headers,
+                                                 ssl=ssl_context)
+        except websockets.exceptions.InvalidStatusCode as e:
+            if e.status_code == 401:
+                e.args = [
+                    f"{self.host} requires BasicAuth credentials. Please either include that argument or check the validity. Websocket Error: {e.args}"]
+                raise(e)
+            elif e.status_code == 403:
+                e.args = [
+                    f"Gateway Token is Invalid: {self.gateway_token} Websocket Error: {e.args}"]
+                raise(e)
+            elif e.status_code == 404:
+                logger.warning("Received 404 when trying to connect, retrying.")
+                self.websocket = None
```

Comment:
I don't see where this sleeps to prevent an immediate retry. In fact, where does the retry happen? `connect_with_retries` looks like it'll return.

Reply:
I've been unable to successfully test this, so I don't know if it'll work.

Reply:
Try making the end of the URL wrong, like staging.majortom.cloud/foo?

Reply:
I changed the logic here now, so it should catch it appropriately.